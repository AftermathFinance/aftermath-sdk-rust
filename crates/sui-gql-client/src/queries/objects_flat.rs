use cynic::QueryFragment;
use serde::Deserialize;

use super::fragments::{ObjectFilter, PageInfoForward};
use crate::{schema, Paged};

/// Paged object data where the `Object` fragment does take any parameters.
#[derive(cynic::QueryFragment, Clone, Debug)]
#[cynic(variables = "Variables")]
pub struct Query<Object>
where
    Object: QueryFragment<SchemaType = schema::Object, VariablesFields = ()>,
{
    #[arguments(filter: $filter, first: $first, after: $after)]
    pub objects: ObjectConnection<Object>,
}

impl<Object> Query<Object> where
    Object: QueryFragment<SchemaType = schema::Object, VariablesFields = ()>
{
}

impl<Object> Paged for Query<Object>
where
    Object: for<'de> Deserialize<'de>
        + QueryFragment<SchemaType = schema::Object, VariablesFields = ()>
        + Send,
{
    type Input = Variables;

    type NextInput = Variables;

    type NextPage = Self;

    fn next_variables(&self, mut prev_vars: Self::Input) -> Option<Self::NextInput> {
        let PageInfoForward {
            has_next_page,
            end_cursor,
        } = &self.objects.page_info;
        if *has_next_page {
            prev_vars.after.clone_from(end_cursor);
            Some(prev_vars)
        } else {
            None
        }
    }
}

// =============================================================================
//  Variables
// =============================================================================

#[derive(cynic::QueryVariables, Clone, Debug)]
pub struct Variables {
    pub filter: Option<ObjectFilter>,
    pub after: Option<String>,
    pub first: Option<i32>,
}

// =============================================================================
//  Inner query fragments
// =============================================================================

/// `ObjectConnection` where the `Object` fragment does take any parameters.
#[derive(cynic::QueryFragment, Clone, Debug)]
#[cynic(graphql_type = "ObjectConnection")]
pub struct ObjectConnection<Object>
where
    Object: QueryFragment<SchemaType = schema::Object, VariablesFields = ()>,
{
    pub nodes: Vec<Object>,
    pub page_info: PageInfoForward,
}
