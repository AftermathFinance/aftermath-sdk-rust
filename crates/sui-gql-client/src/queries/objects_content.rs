use std::collections::HashMap;

use af_sui_types::ObjectId;
use sui_gql_schema::schema;

use super::fragments::{MoveValueRaw, ObjectFilter};
use super::objects_flat::Variables;
use super::outputs::RawMoveStruct;
use super::{objects_flat, Error};
use crate::{missing_data, GraphQlClient};

pub async fn query<C: GraphQlClient>(
    client: &C,
    object_ids: Vec<ObjectId>,
) -> Result<HashMap<ObjectId, RawMoveStruct>, Error<C::Error>> {
    let vars = Variables {
        after: None,
        first: None,
        filter: Some(ObjectFilter {
            object_ids: Some(object_ids),
            ..Default::default()
        }),
    };
    let (init, mut pages) = client
        .query_paged::<Query>(vars)
        .await
        .map_err(Error::Client)?
        .try_into_data()?
        .ok_or(missing_data!("Empty response data"))?;
    pages.insert(0, init);

    let mut raw_objs = HashMap::new();
    for object in pages.into_iter().flat_map(|q| q.objects.nodes) {
        let object_id = object.object_id;
        let struct_ = object
            .as_move_object
            .ok_or(missing_data!("Not a Move object"))?
            .contents
            .ok_or(missing_data!("Object contents"))?
            .try_into()
            .expect("Only structs can be top-level objects");
        raw_objs.insert(object_id, struct_);
    }
    Ok(raw_objs)
}

type Query = objects_flat::Query<Object>;

#[cfg(test)]
#[allow(clippy::unwrap_used)]
#[test]
fn gql_output() {
    use cynic::QueryBuilder as _;

    let vars = Variables {
        filter: None,
        first: None,
        after: None,
    };
    let operation = Query::build(vars);
    insta::assert_snapshot!(operation.query);
}

#[derive(cynic::QueryFragment, Clone, Debug)]
struct Object {
    #[cynic(rename = "address")]
    object_id: ObjectId,
    as_move_object: Option<super::fragments::MoveObjectContent<MoveValueRaw>>,
}
