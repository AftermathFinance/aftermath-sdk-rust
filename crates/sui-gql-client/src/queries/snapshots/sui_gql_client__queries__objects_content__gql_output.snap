---
source: crates/sui-gql-client/src/queries/objects_content.rs
expression: operation.query
---
query Query($filter: ObjectFilter, $after: String, $first: Int) {
  objects(filter: $filter, first: $first, after: $after) {
    nodes {
      address
      asMoveObject {
        contents {
          type {
            repr
          }
          bcs
        }
      }
    }
    pageInfo {
      hasNextPage
      endCursor
    }
  }
}
